packages/flutter/lib/src/material/theme_data.dart

```diff
@@ -1634,3 +1671,152 @@ class _FifoCache<K, V> {
+/// "compactness" of the elements in the UI. It is unitless, since it means
+/// different things to different UI elements.
+///
+/// For example, for buttons, it affects the spacing around the child of the
+/// button. For lists, it affects the distance between baselines of entries in
+/// the list. For chips, it only affects the vertical size, not the horizontal
+/// size.
+class VisualDensity extends Diagnosticable {
+  /// A const constructor for [VisualDensity].
+  ///
+  /// All of the arguments must be non-null, and [horizontal] and [vertical]
+  /// must be in the interval between [minimumDensity] and [maximumDensity],
+  /// inclusive.
+  const VisualDensity({
+    this.horizontal = 0.0,
+    this.vertical = 0.0,
+    this.interval = 4.0,
+  })  : assert(horizontal != null),
+        assert(vertical != null),
+        assert(interval != null),
+        assert(vertical <= maximumDensity),
+        assert(vertical >= minimumDensity),
+        assert(horizontal <= maximumDensity),
+        assert(horizontal >= minimumDensity);
+
+
+  /// The minimum allowed density.
+  static const double minimumDensity = -4.0;
+
+  /// The maximum allowed density.
+  static const double maximumDensity = 4.0;
+
+
+  /// The default profile for [VisualDensity] in [ThemeData].
+  static const VisualDensity standard = VisualDensity();
+
+  /// The profile for a "comfortable" interpretation of [VisualDensity].
+  ///
+  /// Components will modify the base density with this profile in mind, making
+  /// themselves more or less dense based on the Material Design specification
+  /// for this profile for their particular use case.
+  static const VisualDensity comfortable = VisualDensity(horizontal: -1.0, vertical: -1.0);
+
+  /// The profile for a "compact" interpretation of [VisualDensity].
+  ///
+  /// Components will modify the base density with this profile in mind, making
+  /// themselves more or less dense based on the Material Design specification
+  /// for this profile for their particular use case.
+  static const VisualDensity compact = VisualDensity(horizontal: -2.0, vertical: -2.0);
+
+  /// Copy the current [VisualDensity] with the given values replacing the
+  /// current values.
+  VisualDensity copyWith({
+    double horizontal,
+    double vertical,
+    double interval,
+  }) {
+    return VisualDensity(
+      horizontal: horizontal ?? this.horizontal,
+      vertical: vertical ?? this.vertical,
+      interval: interval ?? this.interval,
+    );
+  }
+
+  /// The horizontal visual density of UI components.
+  ///
+  /// This property affects only the horizontal spacing between and within
+  /// components, to allow for different UI visual densities.
+  ///
+  /// See also:
+  ///
+  ///  - [ThemeData.density], where this property is used to specify the base
+  ///    horizontal density of Material components.
+  ///  - [ButtonThemeData.density], where this property is used to specify the
```

Comment:
I think we're leaving density out of the component themes.

Here and elsewhere

Reply:
Done.